arkane/ess/factory.py

```diff
@@ -0,0 +1,251 @@
+###############################################################################
+
+"""
+A module for generating ESS adapters
+"""
+
+import logging
+import os
+from typing import Type
+
+from .adapter import ESSAdapter
+
+from rmgpy.exceptions import InputError
+from rmgpy.qm.qmdata import QMData
+from rmgpy.qm.symmetry import PointGroupCalculator
+
+_registered_ess_adapters = {}
+
+
+def register_ess_adapter(ess: str,
+                         ess_class: Type[ESSAdapter],
+                         ) -> None:
+    """
+    A register for the ESS adapters.
+
+    Args:
+        ess: A string representation for an ESS adapter
+        ess_class: The ESS adapter class
+
+    Raises:
+        TypeError: If ``ess_class`` is not an ``ESSAdapter`` instance.
+    """
+    if not issubclass(ess_class, ESSAdapter):
+        raise TypeError(f'{ess_class} is not an ESSAdapter')
+    _registered_ess_adapters[ess] = ess_class
+
+
+def ess_factory(fullpath: str) -> Type[ESSAdapter]:
+    """
+    A factory generating the ESS adapter corresponding to ``ess_adapter``.
+    Given a path to the log file of a QM software, determine whether it is
+    Gaussian, Molpro, QChem, Orca, or TeraChem
+
+    Args:
+        fullpath (str): The disk location of the output file of interest.
+
+    Returns:
+        Type[ESSAdapter]: The requested ESSAdapter child, initialized with the respective arguments.
+    """
+
+    ess_name = None
+    with open(fullpath, 'r') as f:
+        if os.path.splitext(fullpath)[1] in ['.xyz', '.dat', '.geometry']:
+            ess_name = 'terachem'
+        line = f.readline()
+        while ess_name is None and line != '':
+            if 'gaussian' in line.lower():
+                ess_name = 'gaussian'
+                break
+            elif 'molpro' in line.lower():
+                ess_name = 'molpro'
+                break
+            elif 'O   R   C   A' in line or 'orca' in line.lower():
+                ess_name = 'orca'
+                break
+            elif 'qchem' in line.lower():
+                ess_name = 'qchem'
+                break
+            elif 'terachem' in line.lower():
+                ess_name = 'terachem'
+                break
+            line = f.readline()
+    if ess_name is None:
+        raise InputError(f'The file at {fullpath} could not be identified as a '
+                         f'Gaussian, Molpro, Orca, QChem, or TeraChem log file.')
+
+    return _registered_ess_adapters[ess_name](path=fullpath)
+
+
+class Log(object):
+    def __init__(self, fullpath: str):
+        """
+        This class is not a part of the Factory/Adapter design class. It was added here to keep Arkane
+        backward compatible with current input files that use ``Log`` calls. Added a .ess attribute that stores
+        the specific ESS Adapter, along with shortcut methods to call the respective Adapter method.
+        """
+        self._path = fullpath
+        self.ess = None
+        self.set_ess()
```

Comment:
I like that we are making sure that we don't break backwards compatibility, but I don't want us to have to maintain both the new adapter/factory setup and this Log setup. Can we just have a Log object call the factory and return the adapter? Are there other complications that I am not thinking of here?

Reply:
in some cases the fullpath passed to Log is incomplete. Arkane statmech then troubleshoots it, but since Log is defined in the input file, it must accept a path, but cannot return the Adapter in such cases

Reply:
Ahh, I see. I think we can get around this problem by doing the following.

1. Delete the `Log` class completely
2. In line 233 or so in statmech.py (after `directory` is defined) add the `create_log` function described below.
```
        def create_log(log_path):
            if not os.path.isfile(log_path):
                modified_log_path = os.path.join(directory, log_path)
                if not os.path.isfile(modified_log_path):
                    raise InputError('Could not find single point energy file for species {0} '
                                     'in the specified path {1}'.format(self.species.label, log_path))
                else:
                    log_path = modified_log_path
            
            return ess_factory(log_path)
```

3. Remove the ess_factory code around line 339, as it has been moved inside of our function. Remove any other remaining `isinstance` checks for `Log`.
4. Remove the now redundant path debugging lines in statmech.py around 320
5. Modify the `local_context` definition as follows:
```
        local_context = {
            '__builtins__': None,
            'True': True,
            'False': False,
            'HinderedRotor': hinderedRotor,
            'FreeRotor': freeRotor,
            'HinderedRotor2D': hinderedRotor2D,
            'HinderedRotorClassicalND': hinderedRotorClassicalND,
            'ScanLog': ScanLog,
            'Log': create_log  # The log class no longer exists, so route the path to ess_factory instead
        }

        log_classes = ['GaussianLog', 'MolproLog', 'OrcaLog', 'QChemLog', 'TeraChemLog']
        local_context.update({ess_adapter_name: create_log for ess_adapter_name in log_classes})
```
I wanted to use `_registered_ess_adapters.keys()` to automate that last part, but I realize that the keys don't match exactly with what we need here.

This solution gets rid of the log class and handles the path debugging. @kspieks let me know if you think this will work. If so I am happy to help put this together. If there is a better solution that you prefer let me know as well. Also feel free to add tweaks to the code above if I accidentally left out important checks or such.

Reply:
On its face this looks like a perfect fix! Seems very slick! I'm glad you suggested it. I'll implement this tomorrow morning and make sure all unit tests pass so that it's ready to go before RMG study group :)

Reply:
Implemented the changes and am running unit tests now. I changed the keys in `_registered_ess_adapters.keys()` to be `'GaussianLog'`, rather than `'gaussian'` for example, so that we can say `log_classes = _registered_ess_adapters.keys()` to automate your 5th suggestion above

Reply:
Sounds good, thanks! Just make sure to update `ess_factory` or any other code that was expecting the keys to be something like `gaussian`